Cesium3DTiles/src/RasterizedPolygonsOverlay.cpp

```diff
@@ -0,0 +1,222 @@
+
+#include "Cesium3DTiles/RasterizedPolygonsOverlay.h"
+#include "Cesium3DTiles/BoundingVolume.h"
+#include "Cesium3DTiles/RasterOverlayTileProvider.h"
+#include "Cesium3DTiles/spdlog-cesium.h"
+#include "CesiumAsync/AsyncSystem.h"
+#include "CesiumAsync/IAssetAccessor.h"
+#include "CesiumGeospatial/GlobeRectangle.h"
+#include "CesiumUtility/IntrusivePointer.h"
+#include <memory>
+#include <string>
+
+namespace Cesium3DTiles {
+namespace {
+void rasterizePolygons(
+    CesiumGltf::ImageCesium& image,
+    const CesiumGeospatial::GlobeRectangle& rectangle,
+    const std::string& textureTargetName,
+    const std::vector<CartographicSelection>& cartographicSelections) {
+
+  double rectangleWidth = rectangle.computeWidth();
+  double rectangleHeight = rectangle.computeHeight();
+
+  // create source image
+  image.width = 256;
+  image.height = 256;
+  image.channels = 1;
+  image.bytesPerChannel = 1;
+  image.pixelData.resize(65536);
+
+  std::memset(image.pixelData.data(), 0, 65536);
+
+  // TODO: this is naive approach, use line-triangle
+  // intersections to rasterize one row at a time
+  // NOTE: also completely ignores antimeridian (really these
+  // calculations should be normalized to the first vertex)
+  for (const CartographicSelection& selection : cartographicSelections) {
+
+    if (selection.getTargetTextureName() != textureTargetName) {
+      continue;
+    }
+
+    const std::vector<glm::dvec2>& vertices = selection.getVertices();
+    const std::vector<uint32_t>& indices = selection.getIndices();
+    for (size_t triangle = 0; triangle < indices.size() / 3; ++triangle) {
+      const glm::dvec2& a = vertices[indices[3 * triangle]];
+      const glm::dvec2& b = vertices[indices[3 * triangle + 1]];
+      const glm::dvec2& c = vertices[indices[3 * triangle + 2]];
+
+      glm::dvec2 ab = b - a;
+      glm::dvec2 ab_perp(-ab.y, ab.x);
+      glm::dvec2 bc = c - b;
+      glm::dvec2 bc_perp(-bc.y, bc.x);
+      glm::dvec2 ca = a - c;
+      glm::dvec2 ca_perp(-ca.y, ca.x);
+
+      for (size_t j = 0; j < 256; ++j) {
```

Comment:
I think it doesn't need to run 256x256 for most of the triangles. Another optimization I can think of is to calculate the intersection between the scan line and and the triangle, then determine which the range of the pixels that need to be rasterized. You can also calculate the vertical range of the pixels that the triangle covers to rasterize instead of rasterizing all row pixels
```
                                              O
                                           O    O
     rasterize this pixel ->  O         O <- to this pixel
                                       OOOOOOO
```

Reply:
the 2D line and triangle intersection hopefully should be fast since the line is also a parallel line